components/director/internal/authenticator/middleware.go

```diff
@@ -24,20 +26,24 @@ const AuthorizationHeaderKey = "Authorization"
 type Authenticator struct {
 	jwksEndpoint        string
 	allowJWTSigningNone bool
-	Jwks                *jwk.Set
+	cachedJWKs          *jwk.Set
+	mux                 sync.Mutex
 }
 
 func New(jwksEndpoint string, allowJWTSigningNone bool) *Authenticator {
 	return &Authenticator{jwksEndpoint: jwksEndpoint, allowJWTSigningNone: allowJWTSigningNone}
 }
 
 func (a *Authenticator) SynchronizeJWKS() error {
+	a.mux.Lock()
+	log.Info("Synchronizing JWKS...")
 	jwks, err := FetchJWK(a.jwksEndpoint)
 	if err != nil {
 		return errors.Wrapf(err, "while fetching JWKS from endpoint %s", a.jwksEndpoint)
 	}
 
-	a.Jwks = jwks
+	a.cachedJWKs = jwks
+	a.mux.Unlock()
```

Comment:
Shouldn't you defer this call right after the Lock? Right now when an error happens you return without calling the Unlock.